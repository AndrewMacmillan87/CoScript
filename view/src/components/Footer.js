import React, { Fragment } from "react";
import { Link, withRouter } from "react-router-dom";

// Redux
import { connect } from "react-redux";
import { getMostRecentProject } from "../redux/actions/documents";

import PropTypes from "prop-types";

import "./footer.css";

export const Footer = ({ user, loading, getMostRecentProject, ...props }) => {
    const authLinks = (
        <div className='footer-links'>
            <Link to='/'>Home</Link>
            <Link to='/profile'>Profile</Link>
            <Link
                to='#'
                onClick={async () => {
                    await getMostRecentProject();
                    const { projectData } = props;
                    projectData
                        ? props.history.push(`/editor${projectData.url}`)
                        : props.history.push("/editor");
                }}>
                Workspace
            </Link>
            <Link to='/projects'>Active Projects</Link>
            <Link to='/publications'>Publications</Link>
        </div>
    );

    const guestLinks = (
        <div className='footer-links'>
            <Link to='/'>Home</Link>
            <Link to='/projects'>Active Projects</Link>
            <Link to='/publications'>Publications</Link>
        </div>
    );

    const isLoggedIn = user ? user.isLoggedIn : false;

    return (
        <footer className='footer'>
            <h2 className='footer-h2'>CoScript</h2>
            <div className='footer-container'>
                <div className='links-container socials'>
                    <p>Socials</p>
                    <div className='footer-links'>
                        <a href='https://github.com/AndrewMacmillan87/coscript'>
                            GitHub
                        </a>
                        <a href='/'>Twitter</a>
                        <a href='/'>Facebook</a>
                    </div>
                </div>
                <div className='links-container site-map'>
                    <p>Site Map</p>

                    {!loading && (
                        <Fragment>
                            {isLoggedIn ? authLinks : guestLinks}
                        </Fragment>
                    )}
                </div>
                <div className='links-container legals'>
                    <p>Legal</p>
                    <div className='footer-links'>
                        <a href='/terms'>Terms and Conditions</a>
                        <a href='/copyright'>Copyright Info</a>
                        <a href='/agreement'>Writers Agreements</a>
                        <a href='/'>Accreditation</a>
                    </div>
                </div>
            </div>
        </footer>
    );
};

Footer.propTypes = {
    user: PropTypes.object,
    loading: PropTypes.bool,
    getMostRecentProject: PropTypes.func.isRequired,
};

const mapProps = (state) => ({
    user: state.auth.user,
    loading: state.auth.loading,
    projectData: state.documents.projectData,
});

export default connect(mapProps, { getMostRecentProject })(withRouter(Footer));
