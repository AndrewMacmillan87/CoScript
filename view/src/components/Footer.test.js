import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router, Link } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../redux/store/reduxStore";

import { Footer } from "./Footer";

describe("Footer component", () => {
    let props, mockGetMostRecentProject;
    beforeEach(() => {
        mockGetMostRecentProject = jest.fn();
        props = {
            user: { isLoggedIn: false },
            loading: false,
            getMostRecentProject: mockGetMostRecentProject,
        };
    });

    afterEach(() => {
        props.user.isLoggedIn = false;
    });

    test("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <Footer {...props} />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <Footer {...props} />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a footer element with a className of 'footer'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Footer {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("footer").length).toEqual(1);
        expect(wrapper.find(".footer").length).toEqual(1);
    });

    it("should render a h2 element", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Footer {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("h2").length).toEqual(1);
    });

    it("should render 10 a elements when user not logged in", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Footer {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("a").length).toEqual(10);
    });

    it("should render 7 div elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Footer {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("div").length).toEqual(7);
    });

    it("should render 3 p elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Footer {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").length).toEqual(3);
    });

    it("should render 3 Link components when user not logged in", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Footer {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(Link).length).toEqual(3);
    });

    it("should render 5 Link components when user logged in", () => {
        props.user.isLoggedIn = true;
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Footer {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(Link).length).toEqual(5);
    });

    it("should call `getMostRecentProject()` when 'Workspace' link is clicked", () => {
        props.user.isLoggedIn = true;
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Footer {...props} />
                </Router>
            </Provider>
        );
        wrapper.find(Link).at(2).simulate("click");
        expect(mockGetMostRecentProject).toBeCalled();
    });
});
