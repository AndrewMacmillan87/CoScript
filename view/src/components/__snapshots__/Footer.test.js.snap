// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component renders correctly 1`] = `
<footer
  className="footer"
>
  <h2
    className="footer-h2"
  >
    CoScript
  </h2>
  <div
    className="footer-container"
  >
    <div
      className="links-container socials"
    >
      <p>
        Socials
      </p>
      <div
        className="footer-links"
      >
        <a
          href="https://github.com/AndrewMacmillan87/coscript"
        >
          GitHub
        </a>
        <a
          href="/"
        >
          Twitter
        </a>
        <a
          href="/"
        >
          Facebook
        </a>
      </div>
    </div>
    <div
      className="links-container site-map"
    >
      <p>
        Site Map
      </p>
      <div
        className="footer-links"
      >
        <a
          href="/"
          onClick={[Function]}
        >
          Home
        </a>
        <a
          href="/projects"
          onClick={[Function]}
        >
          Active Projects
        </a>
        <a
          href="/publications"
          onClick={[Function]}
        >
          Publications
        </a>
      </div>
    </div>
    <div
      className="links-container legals"
    >
      <p>
        Legal
      </p>
      <div
        className="footer-links"
      >
        <a
          href="/terms"
        >
          Terms and Conditions
        </a>
        <a
          href="/copyright"
        >
          Copyright Info
        </a>
        <a
          href="/agreement"
        >
          Writers Agreements
        </a>
        <a
          href="/"
        >
          Accreditation
        </a>
      </div>
    </div>
  </div>
</footer>
`;
